File: day02/solution.rb
require_relative '../lib/base.rb'

class Day02 < AdventOfCode
  def self.take_input(s)
    s.lines.map(&:split)
  end

  ROCK = 0
  PAPER = 1
  SCISSORS = 2

  part(1) do |input|
    rps = {
      'A' => ROCK,
      'B' => PAPER,
      'C' => SCISSORS,
      'X' => ROCK,
      'Y' => PAPER,
      'Z' => SCISSORS,
    }
    input.sum do |a, b|
      them = rps[a]
      me = rps[b]
      score = me + 1
      score += 3 if them == me
      score += 6 if (them + 1) % 3 == me
      score
    end
  end

  part(2) do |input|
    rps = {
      'A' => ROCK,
      'B' => PAPER,
      'C' => SCISSORS,
    }
    input.sum{|a,b|
      them = rps[a]
      score = case b
      when 'X' # lose
        0 + (them - 1) % 3
      when 'Y' # draw
        3 + them
      when 'Z' # win
        6 + (them + 1) % 3
      end
      score + 1
    }
  end
end

Day02.run

File: day02/solution.hs
data BattleOutcome = Win | Lose | Draw

data RockPaperScissors = Rock | Paper | Scissors deriving (Eq)

readRPS :: String -> RockPaperScissors
readRPS "A" = Rock
readRPS "B" = Paper
readRPS "C" = Scissors
readRPS "X" = Rock
readRPS "Y" = Paper
readRPS "Z" = Scissors
readRPS _ = error "Invalid Rock-Paper-Scissors symbol"

readStrategy :: String -> BattleOutcome
readStrategy "X" = Lose
readStrategy "Y" = Draw
readStrategy "Z" = Win
readStrategy _ = error "Invalid strategy"

score :: BattleOutcome -> Int
score Lose = 0
score Draw = 3
score Win = 6

value :: RockPaperScissors -> Int
value Rock     = 1
value Paper    = 2
value Scissors = 3

whatBeats :: RockPaperScissors -> RockPaperScissors
whatBeats Rock     = Paper
whatBeats Paper    = Scissors
whatBeats Scissors = Rock

whatLosesTo :: RockPaperScissors -> RockPaperScissors
whatLosesTo = whatBeats . whatBeats

battle :: RockPaperScissors -> RockPaperScissors -> BattleOutcome
battle a b
  | a == whatBeats b = Win
  | b == whatBeats a = Lose
  | otherwise        = Draw

toTuple :: [a] -> (a, a)
toTuple [a, b] = (a, b)

main = do
    inputStr <- getContents
    let inputLines = lines inputStr

    let scoreBattle (them, i) = do
        case i `battle` them of
            Win  -> 6 + value i
            Draw -> 3 + value i
            Lose -> 0 + value i

    let readBattleStrat (theirPickStr, strat) = do
        case readStrategy strat of
            Win  -> (them, whatBeats them)
            Draw -> (them, them)
            Lose -> (them, whatLosesTo them)
            where them = readRPS theirPickStr

    let readBattlePart1 = toTuple . map readRPS . words
    let readBattlePart2 = readBattleStrat . toTuple . words
    print $ sum $ map (scoreBattle . readBattlePart1) inputLines
    print $ sum $ map (scoreBattle . readBattlePart2) inputLines
